cc/algorithms/partition-selection-strategies_test.cc

```diff
@@ -0,0 +1,373 @@
+ }
+
+//We expect the probability of keeping a partition
+//with no users will always be zero
+TEST(PartitionSelectionStrategiesTest, PreaggPartitionSelectionNoUsers) {
+	PreaggPartitionSelection::Builder test_builder;
+	std::unique_ptr<PartitionSelectionStrategy> build =
+		test_builder.SetEpsilon(0.5).SetDelta(0.02)
+					.SetMaxPartitionsContributed(1).Build().ValueOrDie();
+  	double num_kept = 0.0;
+  	for(int i = 0; i < 100000; i++) {
+  		if(build->ShouldKeep(0))
+  			num_kept++;
+  	}
+  	EXPECT_THAT(num_kept/100000, DoubleEq(0));
+ }
+
+TEST(PartitionSelectionStrategiesTest, PreaggPartitionSelectionFirstCrossover) {
+	 PreaggPartitionSelection::Builder test_builder;
+	std::unique_ptr<PartitionSelectionStrategy> build =
+		test_builder.SetEpsilon(0.5).SetDelta(0.02)
+					.SetMaxPartitionsContributed(1).Build().ValueOrDie();
+	PreaggPartitionSelection* magic =
+		dynamic_cast<PreaggPartitionSelection*>(build.get());
+  	EXPECT_THAT(magic->GetFirstCrossover(), DoubleEq(6));
+ }
+
+
+ TEST(PartitionSelectionStrategiesTest,
+ 	  PreaggPartitionSelectionSecondCrossover) {
+	PreaggPartitionSelection::Builder test_builder;
+	std::unique_ptr<PartitionSelectionStrategy> build =
+		test_builder.SetEpsilon(0.5).SetDelta(0.02)
+					.SetMaxPartitionsContributed(1).Build().ValueOrDie();
+	PreaggPartitionSelection* magic =
+		dynamic_cast<PreaggPartitionSelection*>(build.get());
+  	EXPECT_THAT(magic->GetSecondCrossover(), DoubleEq(11));
+ }
+
+//Values calculated with formula
+TEST(PartitionSelectionStrategiesTest,
+	 PreaggPartitionSelectionNumUsersEqFirstCrossover) {
+	PreaggPartitionSelection::Builder test_builder;
+	std::unique_ptr<PartitionSelectionStrategy> build =
+		test_builder.SetEpsilon(0.5).SetDelta(0.02)
+					.SetMaxPartitionsContributed(1).Build().ValueOrDie();
+  	double num_kept = 0.0;
+  	for(int i = 0; i < 10000000; i++) {
+  		if(build->ShouldKeep(6))
+  			num_kept++;
+  	}
+  	EXPECT_THAT(num_kept/10000000, DoubleNear(0.58840484458, 0.001));
+ }
+
+//Values calculated with formula
+TEST(PartitionSelectionStrategiesTest,
+	 PreaggPartitionSelectionNumUsersBtwnCrossovers) {
+	PreaggPartitionSelection::Builder test_builder;
+	std::unique_ptr<PartitionSelectionStrategy> build =
+		test_builder.SetEpsilon(0.5).SetDelta(0.02)
+					.SetMaxPartitionsContributed(1).Build().ValueOrDie();
+  	double num_kept = 0.0;
+  	for(int i = 0; i < 10000000; i++) {
+  		if(build->ShouldKeep(8))
+  			num_kept++;
+  	}
+  	EXPECT_THAT(num_kept/10000000, DoubleNear(0.86807080625, 0.001));
+ }
+
+//Values calculated with formula
+TEST(PartitionSelectionStrategiesTest,
+	 PreaggPartitionSelectionNumUsersGreaterThanCrossovers) {
+	PreaggPartitionSelection::Builder test_builder;
+	std::unique_ptr<PartitionSelectionStrategy> build =
+		test_builder.SetEpsilon(0.5).SetDelta(0.02)
+					.SetMaxPartitionsContributed(1).Build().ValueOrDie();
+  	double num_kept = 0.0;
+  	for(int i = 0; i < 10000000; i++) {
```

Comment:
Same thing here - use EXPECT_TRUE rather than calculating the probability.

Reply:
Once again, just one sample >__<